File: day16/main.cpp
#include <iostream>
#include <filesystem>
#include <fstream>

auto inPath = "day16/test";

enum Direction {
    RIGHT,
    UP,
    LEFT,
    DOWN
};

struct Pos2d {
    int x, y;
};

struct Grid {
    char* data;
    std::vector<char*> rows;
    int maxRow;
    int maxCol;

    bool inBounds(int col, int row) {
        return col >= 0 && col < maxCol && row >= 0 && row < maxRow;
    }

    char get(int col, int row) {
        return rows[row][col];
    }

    static Grid makeUniform(int maxRow, int maxCol) {
        Grid ret;

        for (int i = 0; i < maxRow; i++) {
            ret.rows.push_back(new char[maxCol]);
        }

        ret.maxRow = maxRow;
        ret.maxCol = maxCol;

        return ret;
    }

    static Grid makeFromStr(char* str) {
        Grid ret;
        ret.data = str;

        for (char* i = str; *i != '\0'; i++) {
            if (*i == '\n') {
                ret.rows.emplace_back(i);
            }
        }

        ret.maxRow = ret.rows.size() - 1;

        return ret;
    }

    int count(char c) {
        int cnt = 0;
        for (int row = 0; row < maxRow; row++) {
            for (int col = 0; col < maxCol; col++) {
                if (get(col, row) == c) {
                    cnt++;
                }
            }
        }
        return cnt;
    }
};

void main() {
    std::ifstream file(inPath);
    auto size = std::filesystem::file_size(inPath);
    std::vector<char> contents(size);
    file.read(contents.data(), contents.size());

    std::vector<char> energized(size);

    Grid g = Grid::makeFromStr(contents.data());
    Grid energized = Grid::makeUniform(g.maxRow, g.maxCol);
}

File: day16/main.py
import re, os, functools, sys
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), os.path.pardir)))
from util import *
from math import gcd
from functools import *

file = open(os.path.dirname(os.path.realpath(__file__)) + '/input')
#file = open(os.path.dirname(os.path.realpath(__file__)) + '/test')

lines = file.readlines()

RIGHT = 0
UP = 90
LEFT = 180
DOWN = 270

nextDir = {
    RIGHT: Pos2d(1, RIGHT),
    UP: Pos2d(0, -1),
    LEFT: Pos2d(-1, RIGHT),
    DOWN: Pos2d(0, 1),
}

beams = [(Pos2d(-1, 0), RIGHT)]

g = Grid.makeFromStrs(lines)
energized = Grid.makeUniform(g.maxRow, g.maxCol, '.')
energized.set(Pos2d(0, 0), '#')

alreadySimulated = set()

while True:
    nextBeamList = []
    for beam in beams:
        nextPos = beam[0] + nextDir[beam[1]]
        dir = beam[1]

        nextCell = g.get(nextPos)

        out1 = ()
        out2 = ()

        if nextCell == '.':
            out1 = (nextPos, dir)
        elif nextCell == '\\':
            if dir == RIGHT:
                out1 = (nextPos, DOWN)
            elif dir == UP:
                out1 = (nextPos, LEFT)
            elif dir == LEFT:
                out1 = (nextPos, UP)
            elif dir == DOWN:
                out1 = (nextPos, RIGHT)
        elif nextCell == '/':
            if dir == RIGHT:
                out1 = (nextPos, UP)
            elif dir == UP:
                out1 = (nextPos, RIGHT)
            elif dir == LEFT:
                out1 = (nextPos, DOWN)
            elif dir == DOWN:
                out1 = (nextPos, LEFT)
        elif nextCell == '|':
            if dir == UP or dir == DOWN:
                out1 = (nextPos, dir)
            else:
                out1 = (nextPos, UP)
                out2 = (nextPos, DOWN)
        elif nextCell == '-':
            if dir == RIGHT or dir == LEFT:
                out1 = (nextPos, dir)
            else:
                out1 = (nextPos, RIGHT)
                out2 = (nextPos, LEFT)

        if out1 != () and g.inBounds(out1[0]) and out1 not in alreadySimulated:
            nextBeamList.append(out1)
            energized.set(out1[0], '#')
            alreadySimulated.add(out1)
        if out2 != () and g.inBounds(out2[0]) and out2 not in alreadySimulated:
            nextBeamList.append(out2)
            energized.set(out2[0], '#')
            alreadySimulated.add(out2)

    #print(energized)
    print(energized.count('#'))
    beams = nextBeamList

print(energized)